pbrd/pbr_vty.c

```diff
@@ -153,325 +128,949 @@ DEFUN_NOSH(no_pbr_map, no_pbr_map_cmd, "no pbr-map PBRMAP [seq (1-700)]",
+ * various configured fields in a rule. So far, these fields are
+ * src/dst IP-address match/set values.
+ *
+ * It is probably possible to perform the same address-family check in
+ * the CLI for single nexthops (set nexthop A.B.C.D|X:X::X:X) but the
+ * address-family is not immediately available for nexthop-groups.
+ * In both the single-nexthop and nexthop-group, the NHT resolution code
+ * sets the "family" field of struct pbr_map_sequence asynchronously.
+ *
+ * There isn't currently any flagging of rules that have a consistent
+ * set of src/dst IP-address match/set values but an asynchronously-resolved
+ * nexthop-group that has a different address-family.
+ *
+ * The match/set IP-address handlers below blindly set "family"; it's
+ * probably possible to wrongly set "family" to, e.g., IPv4 this way after
+ * a v6 NHG has been resolved and break rule removal. It's not clear
+ * how to best address this potential issue.
+ */
+static bool pbr_family_consistent(struct pbr_map_sequence *pbrms,
+				  uint8_t family, uint32_t skip_filter_bm,
+				  uint32_t skip_action_bm, const char **msg)
+{
+	uint32_t filter_bm = pbrms->filter_bm & ~skip_filter_bm;
+	uint32_t action_bm = pbrms->action_bm & ~skip_action_bm;
+
+	if (CHECK_FLAG(filter_bm, PBR_FILTER_SRC_IP) &&
+	    (family != pbrms->src->family)) {
+		if (msg)
+			*msg = "match src-ip";
+		return false;
 	}
-
-	return ret;
+	if (CHECK_FLAG(filter_bm, PBR_FILTER_DST_IP) &&
+	    (family != pbrms->dst->family)) {
+		if (msg)
+			*msg = "match dst-ip";
+		return false;
+	}
+	if (CHECK_FLAG(action_bm, PBR_ACTION_SRC_IP) &&
+	    (family != sockunion_family(&pbrms->action_src))) {
+		if (msg)
+			*msg = "set src-ip";
+		return false;
+	}
+	if (CHECK_FLAG(filter_bm, PBR_ACTION_DST_IP) &&
+	    (family != sockunion_family(&pbrms->action_dst))) {
+		if (msg)
+			*msg = "set dst-ip";
+		return false;
+	}
+	return true;
 }
 
-DEFPY(no_pbr_set_table_range, no_pbr_set_table_range_cmd,
-      "no pbr table range [(10000-4294966272)$lb (10000-4294966272)$ub]",
-      NO_STR
-      PBR_STR
-      "Set table ID range\n"
-      "Set table ID range\n"
-      "Lower bound for table ID range\n"
-      "Upper bound for table ID range\n")
+
+/* clang-format off */
+DEFPY  (pbr_map_match_src,
+	pbr_map_match_src_cmd,
+	"[no] match src-ip ![<A.B.C.D/M|X:X::X:X/M>$prefix]",
+	NO_STR
+	"Match the rest of the command\n"
+	"Choose the src ip or ipv6 prefix to use\n"
+	"v4 Prefix\n"
+	"v6 Prefix\n")
 {
-	pbr_nht_set_tableid_range(PBR_NHT_DEFAULT_LOW_TABLEID,
-				  PBR_NHT_DEFAULT_HIGH_TABLEID);
+	/* clang-format on */
+	struct pbr_map_sequence *pbrms = VTY_GET_CONTEXT(pbr_map_sequence);
+	const char *fmsg = NULL;
+
+	if (!pbrms)
+		return CMD_WARNING_CONFIG_FAILED;
+
+	if (no) {
+		if (!CHECK_FLAG(pbrms->filter_bm, PBR_FILTER_SRC_IP))
+			return CMD_SUCCESS;
+		prefix_free(&pbrms->src);
+		UNSET_FLAG(pbrms->filter_bm, PBR_FILTER_SRC_IP);
+		goto check;
+	}
+
+	assert(prefix);
```

Comment:
The prefix is optional, and we are going to have an assert here if we enter `match src-ip`?

Reply:
Prefix is optional only for the "no" case. Mandatory for non-"no" case.